src/query/api/v1/middleware/rewrite.go

```diff
@@ -124,63 +131,33 @@ func RewriteRangeDuration(
 	if err != nil {
 		return err
 	}
-	startTime, endTime := params.start, params.end
-
-	// Using the prometheus engine in this way should be considered
-	// optional and best effort. Fall back to the frequently accurate logic
-	// of using the start and end time in the request
-	if opts.PrometheusEngineFn != nil {
-		lookback := opts.DefaultLookback
-		if params.isLookbackSet {
-			lookback = params.lookback
-		}
-		engine, err := opts.PrometheusEngineFn(lookback)
-		if err != nil {
-			logger.Debug("Found an error when getting a Prom engine to "+
-				"calculate start/end time for query rewriting. Falling back to request start/end time",
-				zap.String("originalQuery", params.query),
-				zap.Duration("lookbackDuration", lookback))
-		}
-		queryable := fakeQueryable{
-			engine:  engine,
-			instant: opts.Instant,
-		}
-		err = queryable.calculateQueryBounds(params.query, params.start, params.end, fetchOpts.Step)
-		if err != nil {
-			logger.Debug("Found an error when using the Prom engine to "+
-				"calculate start/end time for query rewriting. Falling back to request start/end time",
-				zap.String("originalQuery", params.query))
-		}
-		// calculates the query boundaries in roughly the same way as prometheus
-		startTime, endTime = queryable.getQueryBounds()
-	}
-	attrs, err := store.QueryStorageMetadataAttributes(ctx, startTime, endTime, fetchOpts)
+	// Get the appropriate time range before updating the lookback
+	// This is necessary to cover things like the offset and `@` modifiers.
+	startTime, endTime := getQueryBounds(opts, params, fetchOpts, logger)
+	res, err := findLargestQueryResolution(ctx, store, fetchOpts, startTime, endTime)
 	if err != nil {
 		return err
 	}
-
-	// Find the largest resolution
-	var res time.Duration
-	for _, attr := range attrs {
-		if attr.Resolution > res {
-			res = attr.Resolution
+	updateLookback, updatedLookback := maybeUpdateLookback(params, res, opts)
+	originalLookback := params.lookback
+
+	// We use the lookback as a part of bounds calculation. If the lookback changes, we need to recalculate
+	// If it has changed, we need to recalculate
+	if updateLookback {
+		params.lookback = updatedLookback
+		startTime, endTime = getQueryBounds(opts, params, fetchOpts, logger)
+		res, err = findLargestQueryResolution(ctx, store, fetchOpts, startTime, endTime)
+		if err != nil {
+			return err
 		}
 	}
 
-	// Largest resolution is 0 which means we're routing to the unaggregated namespace.
-	// Unaggregated namespace can service all requests, so return.
-	if res == 0 {
-		return nil
-	}
```

Comment:
Are you sure this `if` is no longer appropriate? Somehow I think in this case we don't want to do any modifications, but I might be missing something.

Reply:
No. You're right. I think I deleted too much when I moved the code around.